File: world_event_log.py
"""world_event_log.py — the host-owned, append-only, hash-chained committed
event log (Horizon Phase 0 Build 3, Phase 2 — spec §13.1-13.2).

THE LOCKED T-Q-HORIZON DECISION, in code: the committed-event log is a
host-owned, append-only, hash-chained store, NOT a Planes list. The value
plane has no `hash` builtin and cannot chain events by content — that is the
whole finding behind the decision. Each event's semantic PAYLOAD (tick,
actor, delta, affected subjects, rationale) is ordinary data derived from a
traced world value; its integrity ENVELOPE (monotonic sequence, previous-
event hash, an opaque authorization receipt, a witnessed `when`) is added
here, entirely outside the language the interpreter evaluates. This module
computes the hash chain in Python, reusing `hashlib.sha256` — the same
primitive `interp.py`'s `_seal` already uses and `js/sha256.mjs` mirrors —
never a language builtin.

Two layers:
  - `append_event`/`verify_chain` — pure functions over a plain list of
    entry dicts, so a host's durably-persisted log can be re-verified after
    reload, independent of any particular `WorldEventLog` instance's
    lifetime (Phase 3's recovery reads a log this way).
  - `WorldEventLog` — the stateful convenience wrapper most callers use: an
    in-memory ordered ledger, appending through the pure functions above and
    forwarding each new entry to the OPTIONAL `host.append_event` (host.py)
    for durability, the same tier `record`/`snapshot` already occupy. Its
    public surface is exactly `append`/`events`/`verify` — no update, no
    delete, no mutate. Undo/redo (§13.3) are appended inverse events plus
    snapshot-and-replay recovery, never destructive history deletion.
"""
import copy
import hashlib

from world_delta import canonical_delta_string

EVENT_LOG_FORMAT_VERSION = 1
GENESIS_HASH = "0" * 64


class WorldEventLogError(Exception):
    """A refusal naming the event-log rule it broke — the same tag/detail/
    fix shape `world_ir.WorldIRError` and `interp.py`'s `PlanesError`
    carry."""

    def __init__(self, tag, detail, fix):
        self.tag = tag
        self.detail = detail
        self.fix = fix
        super().__init__(f"{tag}: {detail}")


def _canonical_entry_string(entry):
    """Deterministic text form of everything in `entry` EXCEPT its own
    `hash` field — the hash is computed FROM this string, so it cannot be
    an input to itself. Reuses `canonical_delta_string` for the payload's
    `delta` sub-object rather than a second serializer."""
    payload = entry["payload"]
    lines = [
        f"format: {entry['format']}",
        f"sequence: {entry['sequence']}",
        f"previous-hash: {entry['previous_hash']}",
        f"when: {entry['when']}",
        f"receipt: {entry['receipt'] if entry['receipt'] is not None else '<none>'}",
        f"tick: {payload['tick']}",
        f"actor: {payload['actor']}",
        f"affected-subjects: {','.join(payload['affected_subjects'])}",
        f"rationale: {payload['rationale']}",
    ]
    if payload["delta"] is None:
        lines.append("delta: <none>")
    else:
        lines.append("delta:")
        lines.extend(f"  {line}" for line in canonical_delta_string(payload["delta"]).splitlines())
    return "\n".join(lines)


def _entry_hash(entry):
    return hashlib.sha256(_canonical_entry_string(entry).encode()).hexdigest()


def append_event(events, payload, when, receipt=None):
    """Pure: given the existing ordered entry list (never mutated), builds
    and returns the NEXT entry — sequence and previous-hash derived from
    `events`' own last element, never from external bookkeeping. The
    caller appends the returned entry; this function never does, which is
    what keeps it usable both by `WorldEventLog.append` and directly by a
    recovery path re-deriving a log's expected next entry."""
    sequence = len(events)
    previous_hash = events[-1]["hash"] if events else GENESIS_HASH
    entry = {
        "format": EVENT_LOG_FORMAT_VERSION,
        "sequence": sequence,
        "previous_hash": previous_hash,
        "when": when,
        "receipt": receipt,
        "payload": payload,
    }
    entry["hash"] = _entry_hash(entry)
    return entry


def verify_chain(events):
    """Recompute the hash chain over `events` (any ordered list of entry
    dicts — not necessarily one a live `WorldEventLog` produced, so a
    host's reloaded, persisted log can be checked too). Returns
    `(True, None)` if every entry's `previous_hash` links to the prior
    entry's own recomputed hash and every entry's own `hash` matches its
    recomputed content; otherwise `(False, index)` naming the first entry
    where the chain breaks — a mutated earlier entry breaks its own slot
    (hash mismatch) or the next one (previous-hash mismatch), whichever
    comes first in sequence order."""
    expected_previous = GENESIS_HASH
    for i, entry in enumerate(events):
        if entry["previous_hash"] != expected_previous:
            return False, i
        recomputed = _entry_hash({k: v for k, v in entry.items() if k != "hash"})
        if recomputed != entry["hash"]:
            return False, i
        expected_previous = entry["hash"]
    return True, None


def events_to_json(events):
    return {"format": EVENT_LOG_FORMAT_VERSION, "events": events}


def events_from_json(doc):
    """The refuse-don't-guess half of the contract, mirroring
    `interp.py`'s `records_from_json`: an unrecognized format version is
    rejected, not silently reinterpreted."""
    version = doc.get("format")
    if version != EVENT_LOG_FORMAT_VERSION:
        raise WorldEventLogError(
            "unrecognized-event-log-format",
            f"event log format {version!r} is not {EVENT_LOG_FORMAT_VERSION}",
            "regenerate the event log with a version of planes matching this "
            "reader's event log format — if the log is newer than what this "
            "reads, upgrade planes instead of regenerating the log")
    return doc["events"]


class WorldEventLog:
    """The append-only ledger. `host` is optional (host.py's
    `append_event`, a no-op by default) — a host that keeps nothing still
    runs every existing test unchanged, the same tier `record`/`snapshot`
    already occupy."""

    def __init__(self, host=None):
        self._host = host
        self._events = []

    def append(self, payload, when, receipt=None):
        entry = append_event(self._events, payload, when, receipt)
        self._events.append(entry)
        if self._host is not None:
            self._host.append_event(entry)
        return entry

    def events(self):
        """A defensive deep copy — the returned list is not a window onto
        `self._events`, so nothing reachable from this call can mutate the
        log. Read-only by construction, not by convention."""
        return copy.deepcopy(self._events)

    def verify(self):
        return verify_chain(self._events)

File: js/world_event_log.mjs
// js/world_event_log.mjs — the host-owned, append-only, hash-chained
// committed event log. Mirrors world_event_log.py field for field; see that
// file's module docstring for the full T-Q-Horizon decision rationale.
//
// Hashing reuses sha256Hex (js/sha256.mjs) — the same primitive
// js/interp.mjs's `_seal` mirror uses — never a language builtin.

import { sha256Hex } from "./sha256.mjs";
import { canonicalDeltaString } from "./world_delta.mjs";

export const EVENT_LOG_FORMAT_VERSION = 1;
export const GENESIS_HASH = "0".repeat(64);

export class WorldEventLogError extends Error {
  constructor(tag, detail, fix) {
    super(`${tag}: ${detail}`);
    this.tag = tag;
    this.detail = detail;
    this.fix = fix;
  }
}

// Deterministic, cross-implementation-stable text form of a plain
// JSON-shaped value: sorted object keys, no whitespace — mirrors Python's
// `json.dumps(value, sort_keys=True, separators=(",", ":"))` exactly (see
// world_event_log.py's own `_canonical_event_string`), so an identical
// logical event produces an identical string, and therefore an identical
// hash, in either implementation regardless of the two languages' own
// differing object/dict key-insertion-order conventions.
function canonicalEventString(value) {
  if (value === null || typeof value !== "object") return JSON.stringify(value);
  if (Array.isArray(value)) return `[${value.map(canonicalEventString).join(",")}]`;
  const keys = Object.keys(value).sort();
  return `{${keys.map((k) => `${JSON.stringify(k)}:${canonicalEventString(value[k])}`).join(",")}}`;
}

// Deterministic text form of everything in `entry` EXCEPT its own `hash`
// field — the hash is computed FROM this string. Reuses
// canonicalDeltaString for the payload's `delta` sub-object.
function canonicalEntryString(entry) {
  const p = entry.payload;
  const lines = [
    `format: ${entry.format}`,
    `sequence: ${entry.sequence}`,
    `previous-hash: ${entry.previousHash}`,
    `when: ${entry.when}`,
    `receipt: ${entry.receipt !== null && entry.receipt !== undefined ? entry.receipt : "<none>"}`,
    `tick: ${p.tick}`,
    `actor: ${p.actor}`,
    `affected-subjects: ${p.affectedSubjects.join(",")}`,
    `rationale: ${p.rationale}`,
  ];
  if (p.delta === null || p.delta === undefined) {
    lines.push("delta: <none>");
  } else {
    lines.push("delta:");
    for (const line of canonicalDeltaString(p.delta).split("\n")) lines.push(`  ${line}`);
  }
  // OPTIONAL, ADDITIVE (Horizon Phase 2 Build 2's own gap-fix). A world-v1
  // caller's payload always carries a real `delta` and never sets `event` —
  // for those payloads this line never appears, so their canonical string
  // and every hash it feeds is BYTE IDENTICAL to before this field existed.
  // The gap this closes: a scene-intent crossing's applied input has no
  // facet-patch delta to log (see main.mjs's own scene-intent path — there
  // is no fold, the program re-emits its full scene every tick), but replay
  // still needs the real typed event object back, not just a human-
  // readable rationale string, to feed the correct per-tick events batch
  // into `advance`. `event`, when present, is arbitrary JSON-shaped data
  // the caller controls the key order of — this is not a general
  // canonicalizer, callers must construct their event objects with a
  // consistent field order (as a_crossing.planes's own UI events already
  // do) for the hash to be reproducible across an identical event.
  if (p.event !== null && p.event !== undefined) {
    lines.push(`event: ${canonicalEventString(p.event)}`);
  }
  return lines.join("\n");
}

export function entryHash(entry) {
  return sha256Hex(canonicalEntryString(entry));
}

// Pure: given the existing ordered entry list (never mutated), builds and
// returns the NEXT entry — sequence and previous-hash derived from
// `events`' own last element.
export function appendEvent(events, payload, when, receipt = null) {
  const sequence = events.length;
  const previousHash = events.length > 0 ? events[events.length - 1].hash : GENESIS_HASH;
  const entry = {
    format: EVENT_LOG_FORMAT_VERSION,
    sequence,
    previousHash,
    when,
    receipt,
    payload,
  };
  entry.hash = entryHash(entry);
  return entry;
}

// Recompute the hash chain over `events` (any ordered list of entry
// objects). Returns [true, null] if consistent, else [false, index] naming
// the first entry where the chain breaks.
export function verifyChain(events) {
  let expectedPrevious = GENESIS_HASH;
  for (let i = 0; i < events.length; i++) {
    const entry = events[i];
    if (entry.previousHash !== expectedPrevious) return [false, i];
    const { hash, ...withoutHash } = entry;
    if (entryHash(withoutHash) !== hash) return [false, i];
    expectedPrevious = entry.hash;
  }
  return [true, null];
}

export function eventsToJson(events) {
  return { format: EVENT_LOG_FORMAT_VERSION, events };
}

export function eventsFromJson(doc) {
  const version = doc.format;
  if (version !== EVENT_LOG_FORMAT_VERSION) {
    throw new WorldEventLogError(
      "unrecognized-event-log-format",
      `event log format ${JSON.stringify(version ?? null)} is not ${EVENT_LOG_FORMAT_VERSION}`,
      "regenerate the event log with a version of planes matching this reader's "
        + "event log format — if the log is newer than what this reads, upgrade "
        + "planes instead of regenerating the log",
    );
  }
  return doc.events;
}

// The append-only ledger. `host` is optional (host.mjs's `appendEvent`, a
// no-op by default) — a host that keeps nothing still runs every existing
// test unchanged, the same tier record/snapshot already occupy.
export class WorldEventLog {
  constructor(host = null) {
    this._host = host;
    this._events = [];
  }

  append(payload, when, receipt = null) {
    const entry = appendEvent(this._events, payload, when, receipt);
    this._events.push(entry);
    if (this._host !== null) this._host.appendEvent(entry);
    return entry;
  }

  // A defensive deep copy — read-only by construction, not by convention.
  events() {
    return structuredClone(this._events);
  }

  verify() {
    return verifyChain(this._events);
  }
}
